:seedling: Add licence-scan for pull requests

killianmuldoon opened the pull request:
This adds a trivy license scan to the Makefile `verify` target which runs on all pull requests. This should allow detection of undesired license changes in go dependencies.

Related to #9181

killianmuldoon commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,32 @@
+
+TRIVY="${REPO_ROOT}/hack/tools/bin/trivy/${VERSION}/trivy"
+
+$TRIVY filesystem . --license-full --ignored-licenses MPL-2.0  --scanners license --exit-code 1 --severity MEDIUM,HIGH,CRITICAL
```
More on license scanning with trivy here: https://aquasecurity.github.io/trivy/v0.44/docs/scanner/license/

I'm ignoring the MPL 2.0 license as we have a number of dependencies with that license. If  we want to more a more advanced config we could create one based on the CNCF license and allowlist.

One disadvantage of the Trivvy scan is that it seems to only scan dependencies in the `go.mod` indirect dependencies from `go.sum` are not scanned it looks like.

sbueringer replied:
This might be intentional. Not everything in go.sum actually makes it into the final binary

killianmuldoon replied:
> ideally, we should converge on the same tool used by K8s later on
I think k/k is using FOSSA which requires an API key to run so would be a higher maintenance burden. I'm not sure if it's much better than trivvy. 

What I meant by the above is that we can configure Trivy to be specific to the CNCF list of allowed licenses and the granted exceptions. There is some maintenance in this too as we'd have to keep the lists in sync.

sbueringer replied:
To be honest I'm struggling a bit between merging something as soon as we can in CAPI vs. exploring a bit more what k/k is doing, comparing it with what we do here and then decide what we should do in CAPI

But maybe I'm just lacking an overview

sbueringer replied:
>One disadvantage of the Trivvy scan is that it seems to only scan dependencies in the go.mod indirect dependencies from go.sum are not scanned it looks like.

This seems to be aligned to k/k?

> I'm ignoring the MPL 2.0 license as we have a number of dependencies with that license

Also in go.mod or only go.sum?

sbueringer replied:
Sounds good, let's do a bit more research and then discuss.

Thx for all the research you're doing on this.

One additional thing I'm struggling with is allowing MPL licences in general while it is not on the allow-list of CNCF. I think more information will help to make a call

sbueringer replied:
For the immediate feature we can be just very conservative with merging anything that bumps a hashicorp dependency or introduces a new dependency

killianmuldoon replied:
> Also in go.mod or only go.sum?

We have one MPL licensed projects in our go.mod after the recent changes:
`github.com/hashicorp/hcl` which is brought in by our usage of the viper library.

sbueringer replied:
> We have one MPL licensed projects in our go.mod after the recent changes:
github.com/hashicorp/hcl which is brought in by our usage of the viper library.

Let's see https://github.com/spf13/viper/issues/1608 :)

killianmuldoon commented:
/area ci

vincepri commented on Makefile:
```diff
@@ -652,7 +654,13 @@ verify-tiltfile: ## Verify Tiltfile format
+verify-licenses: ## Verify licenses
+	TRACE=$(TRACE) ./hack/verify-licenses.sh $(TRIVY_VER)
+
+
```
```suggestion
```

vincepri commented on Makefile:
```diff
@@ -652,7 +654,13 @@ verify-tiltfile: ## Verify Tiltfile format
-	TRACE=$(TRACE) ./hack/verify-container-images.sh
+	TRACE=$(TRACE) ./hack/verify-container-images.sh $(TRIVY_VER)
+
+.PHONY: verify-licenses
+verify-licenses: ## Verify licenses
+	TRACE=$(TRACE) ./hack/verify-licenses.sh $(TRIVY_VER)
```
There is also a github action if we wanted to go that path https://github.com/aquasecurity/trivy-action

killianmuldoon replied:
Good to know - for now though I like the pattern of running all verify jobs at once. And I think there's a general community preference for putting stuff on prow when that's an option.

fabriziopandini replied:
I have considered the trivy action, but running it for every image we were producing wasn't really efficient while now we are pulling trivy + its database only once and reusing it for all the images
But some time is passed, so might be worth re-evaluating

sbueringer replied:
E.g. stuff like k/k seems to ignore go.sum files, they are not accepting MPL licences, ...

killianmuldoon commented:
/cherry-pick release-1.5

killianmuldoon commented:
/cherry-pick release-1.4

furkatgofurov7 commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
```
we can use the up-to-date year perhaps

killianmuldoon replied:
Good catch - fixed now.

furkatgofurov7 commented on hack/ensure-trivy.sh:
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
```
nit: 
```suggestion
# Copyright 2023 The Kubernetes Authors.
```

chrischdi commented on Makefile:
```diff
@@ -652,7 +654,11 @@ verify-tiltfile: ## Verify Tiltfile format
+	TRACE=$(TRACE) ./hack/verify-container-images.sh $(TRIVY_VER)
+
+.PHONY: verify-licenses
+verify-licenses: ## Verify licenses
```
note: I could also take this as a follow-up

sbueringer replied:
Would be a nice follow-up

killianmuldoon commented on CONTRIBUTING.md:
```diff
@@ -415,6 +416,10 @@ There may, at times, need to be exceptions where breaking changes are allowed in
 
+## Dependency Licence Management
+
+Cluster API follows the [license policy of the CNCF](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md). This sets limits on which
```
Very lightweight statement of intention here.

sbueringer replied:
Sounds good!

killianmuldoon commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,40 @@
+TRIVY="${REPO_ROOT}/hack/tools/bin/trivy/${VERSION}/trivy"
+$TRIVY filesystem . --license-full --ignored-licenses ${CNCF_LICENSE_ALLOWLIST} --scanners license --severity UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL -f json | \
+# Specifically ignore 'github.com/hashicorp/hcl'. This is a known indirect dependency that we should remove where possible.
+jq  '.Results[] | select( .Licenses[]?.PkgName != "github.com/hashicorp/hcl")  | if . == {} then .  else error(.) end'
```
Specific ignore for the one non-covered licence  - we have an issue which references and covers this here: https://github.com/kubernetes-sigs/cluster-api/issues/9181

killianmuldoon commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,40 @@
+
+
+TRIVY="${REPO_ROOT}/hack/tools/bin/trivy/${VERSION}/trivy"
+$TRIVY filesystem . --license-full --ignored-licenses ${CNCF_LICENSE_ALLOWLIST} --scanners license --severity UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL -f json | \
```
Ignoring the CNCF allowlist like this means we should get alerts for any introduced license that is not on this list - or an unknown license.

killianmuldoon commented:
Current state:

- Updated this PR to match the exact guidance of the CNCF upstream. 
- Added a specific exclude for `hashicorp/hcl` which we should probably move toward removing. Issue https://github.com/kubernetes-sigs/cluster-api/issues/9181
- Using trivy as it's simpler - i.e. can be run locally with no API - than snyk or fossa which are also used in the community. `k/k` uses a few custom scripts, and we don't have full guidance for how to scan from upstream yet.

sbueringer commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,41 @@
+source "${REPO_ROOT}/hack/ensure-trivy.sh"
+
+
+TRIVY="${REPO_ROOT}/hack/tools/bin/trivy/${VERSION}/trivy"
```
Q: can trivy also scan licences based on image?

killianmuldoon replied:
It can - we could add it to the existing image scans. 

I've been playing around with it, but haven't gotten ideal results. We should probably have a separate issue for image scanning as it's not going to be useful for PRs in any case.

sbueringer commented on hack/verify-container-images.sh:
```diff
@@ -22,48 +22,25 @@ if [[ "${TRACE-0}" == "1" ]]; then
-# Downloads trivy scanner
-curl -L -o ${TOOL_BIN}/trivy.tar.gz "https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_${TRIVY_OS}-${TRIVY_ARCH}.tar.gz"
+REPO_ROOT=$(git rev-parse --show-toplevel)
+source "${REPO_ROOT}/hack/ensure-trivy.sh"
```
How is the version passed into ensure-trivy.sh if we source the script instead of executing it with a parameter?

killianmuldoon replied:
Currently its passed in through the env variable - will update to correctly pass this as a parameter.

sbueringer replied:
@killianmuldoon nit: let's call it the same way in both scripts

sbueringer commented on hack/verify-licenses.sh:
```diff
@@ -0,0 +1,41 @@
+$TRIVY filesystem . --license-full --ignored-licenses ${CNCF_LICENSE_ALLOWLIST} --scanners license --severity UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL -f json | \
+# Specifically ignore 'github.com/hashicorp/hcl'. This is a known indirect dependency that we should remove where possible.
+# This query ensures we only skip github.com/hashicorp/hcl for as long as the license remains MPL-2.0
+jq  '.Results[] | select( .Licenses[]?.PkgName == "github.com/hashicorp/hcl" and .Licenses[]?.Name == "MPL-2.0" | not) | if . == {} then .  else error(.) end'
```
Can we already open an issue to cover getting rid of this dependency and link the issue here in the comment?

killianmuldoon replied:
Will add a note to that issue to cover this over.

sbueringer reviewed (commented):
Nice work. Thank you very much for working on this!

sbueringer commented:
Thank you!

/approve

/assign @chrischdi

chrischdi reviewed (commented):
/lgtm

sbueringer commented on hack/verify-container-images.sh:
```diff
@@ -22,48 +22,25 @@ if [[ "${TRACE-0}" == "1" ]]; then
 
-TRIVY_VERSION=0.34.0
+VERSION=${1}
 
```
@killianmuldoon Found in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2299/files. We probably still want to set GOARCH
